### PR TITLE
Master purchase stock jem

### PR DESCRIPTION
**Pull request by jem-odoo**
Task: https://www.odoo.com/web?debug#id=47927&view_type=form&model=project.task&action=333&active_id=965&menu_id=4720

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/purchase/models/product.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    _name = 'product.product'
+    _inherit = 'product.product'
+
+    @api.multi
+    def _purchase_count(self):
+        domain = [
+            ('state', 'in', ['purchase', 'done']),
+            ('product_id', 'in', self.mapped('id')),
+        ]
+        PurchaseOrderLines = self.env['purchase.order.line'].search(domain)
```
Comment (fmdl) [inline addons/purchase/models/product.py]: Please find an improvement fot this compute : https://github.com/odoo/odoo/pull/23101

File: addons/purchase/models/res_config_settings.py [hunk 2]
```diff
@@ -43,11 +39,12 @@ def _onchange_use_po_lead(self):
     def get_values(self):
         res = super(ResConfigSettings, self).get_values()
         res.update(
-            is_installed_sale=self.env['ir.module.module'].search([('name', '=', 'sale'), ('state', '=', 'installed')]).id,
+            use_po_lead=self.env['ir.config_parameter'].sudo().get_param('purchase.use_po_lead')
```
[the remark this answers is not in the document]
Comment (jem-odoo) [reply addons/purchase/models/res_config_settings.py]: Nope, rebase mistake. Good catch !

File: addons/purchase/models/purchase.py [hunk 3]
```diff
@@ -524,27 +404,6 @@ def _render_note_exception_quantity_po(order_exceptions):
             filtered_documents[(parent, responsible)] = rendering_context
         self.env['stock.picking']._log_activity(_render_note_exception_quantity_po, filtered_documents)
 
-    @api.multi
```
Comment (pimodoo) [inline addons/purchase/models/purchase.py]: Hello, the method _log_decrease_ordered_quantity just above and its sub method seems to have some references to stock pickings and moves

**Comment by pimodoo** [comment]
In file purchase_template.xml, line 37-40, ref to impacted picking set in the other method (all those stuff must come from a forward port maybe)

File: addons/purchase/models/purchase.py [hunk 4]
```diff
@@ -382,27 +317,6 @@ def button_confirm(self):
 
     @api.multi
     def button_cancel(self):
-        for order in self:
-            for pick in order.picking_ids:
-                if pick.state == 'done':
-                    raise UserError(_('Unable to cancel purchase order %s as some receptions have already been done.') % (order.name))
-            for inv in order.invoice_ids:
```
Comment (amoyaux) [inline addons/purchase/models/purchase.py]: @jem-odoo not sure that you should remove the check on invoice

File: addons/purchase/models/purchase.py [hunk 5]
```diff
@@ -612,33 +421,6 @@ def _compute_qty_invoiced(self):
                         qty -= inv_line.uom_id._compute_quantity(inv_line.quantity, line.product_uom)
             line.qty_invoiced = qty
 
-    @api.model
-    def create(self, values):
-        line = super(PurchaseOrderLine, self).create(values)
-        if line.order_id.state == 'purchase':
```
Comment (amoyaux) [inline addons/purchase/models/purchase.py]: @jem-odoo Maybe, this information could be also useful for service. You could maybe keep it. As you prefer for me.

File: addons/purchase/models/purchase.py [hunk 6]
```diff
@@ -612,33 +421,6 @@ def _compute_qty_invoiced(self):
...
-            line._create_or_update_picking()
-            msg = _("Extra line with %s ") % (line.product_id.display_name,)
-            line.order_id.message_post(body=msg)
-        return line
-
-    @api.multi
-    def write(self, values):
-        if 'product_qty' in values:
-            for line in self:
-                if line.order_id.state == 'purchase':
```
Comment (amoyaux) [inline addons/purchase/models/purchase.py]: @jem-odoo Same here, however I'm not sure if it useful for service

File: addons/purchase/models/purchase.py (hunk 4 above)
Comment (jem-odoo) [reply addons/purchase/models/purchase.py]: You are right; I'll keep it in purchase module.

File: addons/purchase/models/purchase.py (hunk 5 above)
Comment (jem-odoo) [reply addons/purchase/models/purchase.py]: Good idea, for sale order, the message_post is done in sale module. SO I'll move this to purchase.

File: addons/purchase/report/purchase_order_templates.xml [hunk 7]
```diff
@@ -6,29 +6,22 @@
...
-                    </div>
                 </div>
                 <div class="col-xs-5 col-xs-offset-1">
                     <div t-field="o.partner_id"
                         t-options='{"widget": "contact", "fields": ["address", "name", "phone"], "no_marker": True, "phone_icons": True}'/>
                         <p t-if="o.partner_id.vat"><t t-esc="o.company_id.country_id.vat_label or 'TIN'"/>: <span t-field="o.partner_id.vat"/></p>
                 </div>
             </div>
+
+            <h2 t-if="o.state != 'draft'">Purchase Order Confirmation #<span t-field="o.name"/></h2>
```
[the remark this answers is not in the document]
Comment (jem-odoo) [reply addons/purchase/report/purchase_order_templates.xml]: Bad rebase, truc de fou !

**Pull request merged**